### PR TITLE
Update SQLite subcommand to use links

### PR DESCRIPTION
**Pull request by kate-goldenring**
PR for managing databases and links.

### CODE REVIEW

File: src/commands/link.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#[clap(about = "Manage links between apps and resources")]
+pub enum LinkCommand {
+    Sqlite(SqliteLinkCommand),
+}
+
+#[derive(Parser, Debug)]
+pub struct SqliteLinkCommand {
+    #[clap(flatten)]
+    common: CommonArgs,
+    link: String,
```
Comment (rylev) [inline src/commands/link.rs:18]: I think we're going to want to make sure that link meets some limits of what we want to support for link names (start with ascii only letters and numbers along with "-" and "_"?)

File: src/commands/sqlite.rs [hunk 2]
```diff
@@ -113,32 +129,89 @@ impl SqliteCommand {
...
+        link: Link,
+    }
+    let no_link_dbs: Vec<_> = databases.iter().filter(|db| db.links.is_empty()).collect();
+    let mut links: Vec<DBLink> = databases
+        .iter()
+        .flat_map(|db| {
+            db.links.iter().map(|l| DBLink {
+                name: db.name.clone(),
+                link: l.clone(),
+            })
```
Comment (rylev) [inline src/commands/sqlite.rs:162]:
```suggestion
            db.links.into_iter().map(|l| DBLink {
                name: db.name.clone(),
                link: l,
            })
```

**Review by rylev — COMMENTED** [review]
Looking great!

File: src/commands/deploy.rs [hunk 3]
```diff
@@ -555,23 +588,91 @@ fn validate_cloud_app(app: &RawAppManifest) -> Result<()> {
     Ok(())
 }
 
-async fn create_default_database_if_does_not_exist(
+async fn prompt_create_link_if_does_not_exist(
     app_name: &str,
-    app_id: Uuid,
+    app_id: Option<Uuid>,
```
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply src/commands/deploy.rs:593]: I did it to enable prompting to just create a database without creating a link. But this makes functions multipurpose to the point of lack of purpose. I agree it should not be optional and that we should create different functions for just creating a database

File: src/commands/deploy.rs [hunk 4]
```diff
@@ -555,23 +588,91 @@ fn validate_cloud_app(app: &RawAppManifest) -> Result<()> {
...
     client: &CloudClient,
-) -> Result<()> {
-    let default_db = client
-        .get_databases(Some(app_id))
-        .await?
-        .into_iter()
-        .find(|d| d.default);
+    link: &str,
+) -> Result<Option<String>> {
+    let databases = client.get_databases(app_id).await?;
```
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply src/commands/deploy.rs:597]: That would be helpful for sure

File: src/commands/deploy.rs [hunk 5]
```diff
@@ -555,23 +588,91 @@ fn validate_cloud_app(app: &RawAppManifest) -> Result<()> {
...
+}
+
+async fn prompt_link_to_new_database(
+    client: &CloudClient,
+    app_name: &str,
+    app_id: Option<Uuid>,
+    link: &str,
+) -> Result<Option<String>> {
+    // TODO: decide what to do for default db names
+    let default_name = format!("{app_name}-db");
```
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply src/commands/deploy.rs:667]: agree. this is a place holder but will update the comment

File: src/commands/deploy.rs [hunk 6]
```diff
@@ -555,23 +588,91 @@ fn validate_cloud_app(app: &RawAppManifest) -> Result<()> {
...
+    let prompt = "What would you like to name your database?".to_string();
+    let name = dialoguer::Input::new()
+        .with_prompt(prompt)
+        .default(default_name)
+        .interact_text()?;
+    let create_link = app_id.map(|_| link.to_string());
```
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply src/commands/deploy.rs:673]: I am trying to convey that if there is an associated app with the request, then we can create a link. Per your previous comment, we should probably split these flows and make appid required here

File: src/commands/link.rs (hunk 1 above)
Comment (kate-goldenring) [reply src/commands/link.rs:18]: agreed. addded a todo

File: src/commands/link.rs [hunk 7]
```diff
@@ -0,0 +1,118 @@
+use crate::commands::create_cloud_client;
+use crate::opts::*;
+use anyhow::{Context, Result};
+use clap::{Args, Parser};
+use cloud::{client::Client as CloudClient, mocks::Database};
+
+/// Manage links between apps and resources
+#[derive(Parser, Debug)]
+#[clap(about = "Manage links between apps and resources")]
```
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply src/commands/link.rs:9]: Yeah this is redundant

**Comment by rylev** [comment]
I'm going to mark this as ready to review, because it works well already.

File: crates/cloud/src/client.rs [hunk 8]
```diff
@@ -131,18 +131,13 @@ impl Client {
...
-        create_default_database: bool,
-    ) -> Result<Uuid> {
+    pub async fn add_app(&self, name: &str, storage_id: &str) -> Result<Uuid> {
         api_apps_post(
             &self.configuration,
             CreateAppCommand {
                 name: name.to_string(),
                 storage_id: storage_id.to_string(),
-                create_default_database: Some(create_default_database),
+                create_default_database: None,
```
[the remark this answers is not in the document]
Comment (rylev) [reply crates/cloud/src/client.rs:141]: Thinking about this - the answer is: once we no longer support versions of the cloud plugin that use this option.

**Review by rylev — APPROVED** [review]

**Comment by kate-goldenring** [comment]
@rylev this is a lot of commits but rebasing was a headache. If you aren't opposed, i copied all changes to a different branch. I could push that branch into this one to get us down to one commit: https://github.com/fermyon/cloud-plugin/compare/main...kate-goldenring:sql-links-cli?expand=1

**Comment by rylev** [comment]
That might be easiest... I wouldn't be opposed.

**Comment by kate-goldenring** [comment]
@rylev done 🎉

File: src/commands/deploy.rs [hunk 9]
```diff
@@ -614,32 +583,202 @@ fn validate_cloud_app(app: &RawAppManifest) -> Result<()> {
...
+        .interact_opt()?
+    {
+        Some(i) => i,
+        None => return Ok(DatabaseSelection::Cancelled),
+    };
+    Ok(DatabaseSelection::Existing(database_names.remove(index)))
+}
+
+fn prompt_link_to_new_database(name: &str, label: &str) -> Result<DatabaseSelection> {
+    // TODO: use random name generator
```
Comment (kate-goldenring) [inline src/commands/deploy.rs:676]: @rylev do we want to use app name or generate a random name as we have been?

**Comment by kate-goldenring** [comment]
@rylev i am putting this back in draft because there are a few things we should figure out before merging:
1. What is the default DB name
2. We should improve the deploy flow for when a user cancels db selection during a deploy. It breaks future attempts because the oci image is pushed and storage id is set:
```sh
Uploading todo-three version 0.1.1-re3ef23e to Fermyon Cloud...
Deploying...
App "todo-three" accesses a database labeled "default"
Would you like to link an existing database or create a new database?: Create a new database and link the app to it
What would you like to name your database?
Note: This name is used when managing your database at the account level. The app "todo-three" will refer to this database by the label "default".
Other apps can use different labels to refer to the same database.: todo-three-db
Error: Unable to create app

Caused by:
    {"StorageId":["Storage ID not found in storage service"]}

Learn more at https://developer.fermyon.com/cloud/faq
```

**Comment by rylev** [comment]
Talking it over with others, we've decided to go with random name generation for now. We can always revisit this in the future.

**Pull request merged**